File: internal/ogamed/retry.go
package ogamed

import (
	"context"
	"fmt"
	"math"
	"math/rand"
	"time"

	"log/slog"
)

// RetryConfig controls the retry behavior for transient failures.
type RetryConfig struct {
	MaxAttempts  int     // default 3
	BaseDelayMs  int     // default 1000
	MaxDelayMs   int     // default 30000
	JitterFactor float64 // default 0.25 (±25% jitter)
}

// DefaultRetryConfig provides sensible defaults for retry behavior.
var DefaultRetryConfig = RetryConfig{
	MaxAttempts:  3,
	BaseDelayMs:  1000,
	MaxDelayMs:   30000,
	JitterFactor: 0.25,
}

// IsRetryable determines if an error should be retried.
// Non-retryable: 4xx client errors.
// Retryable: network errors, 5xx server errors.
var IsRetryable = func(err error) bool {
	if ogamedErr, ok := err.(*OgamedError); ok {
		return ogamedErr.Code >= 500 || ogamedErr.Code == 0
	}
	// Network errors and other transient failures are retryable
	return true
}

// retryWithBackoff executes fn with exponential backoff on transient failures.
// Non-retryable errors (e.g., 4xx client errors) are returned immediately.
func retryWithBackoff(ctx context.Context, fn func() error, cfg RetryConfig, log *slog.Logger) error {
	for attempt := 1; attempt <= cfg.MaxAttempts; attempt++ {
		err := fn()
		if err == nil {
			return nil
		}

		if !IsRetryable(err) {
			return err
		}

		if attempt >= cfg.MaxAttempts {
			return err
		}

		// Exponential backoff with jitter
		baseDelay := math.Min(float64(cfg.BaseDelayMs)*math.Pow(2, float64(attempt-1)), float64(cfg.MaxDelayMs))
		jitter := baseDelay * cfg.JitterFactor * (rand.Float64()*2 - 1)
		delay := time.Duration(math.Max(0, baseDelay+jitter)) * time.Millisecond

		log.Warn("Request failed, retrying",
			"attempt", attempt,
			"maxAttempts", cfg.MaxAttempts,
			"delay_ms", delay.Milliseconds(),
			"error", err,
		)

		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(delay):
		}
	}
	return fmt.Errorf("retry loop exited unexpectedly")
}

File: packages/bot/src/client/retry.ts
import type { Logger } from 'pino';
import { OgamedError } from '@ogame-bot/shared';

export interface RetryConfig {
  maxAttempts: number; // default 3
  baseDelayMs: number; // default 1000
  maxDelayMs: number; // default 30000
  jitterFactor: number; // default 0.25 (±25% jitter)
}

const DEFAULT_RETRY_CONFIG: RetryConfig = {
  maxAttempts: 3,
  baseDelayMs: 1000,
  maxDelayMs: 30000,
  jitterFactor: 0.25,
};

export async function retryWithBackoff<T>(
  fn: () => Promise<T>,
  config: Partial<RetryConfig> = {},
  log?: Logger,
): Promise<T> {
  const opts = { ...DEFAULT_RETRY_CONFIG, ...config };

  for (let attempt = 1; attempt <= opts.maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (error) {
      // Don't retry on Zod validation errors or 4xx client errors
      if (error instanceof OgamedError && error.code >= 400 && error.code < 500) {
        throw error;
      }
      // Don't retry on ZodError (validation failure = bad data, retrying won't help)
      if (error instanceof Error && error.name === 'ZodError') {
        throw error;
      }

      if (attempt >= opts.maxAttempts) {
        throw error;
      }

      // Exponential backoff with jitter
      const baseDelay = Math.min(opts.baseDelayMs * Math.pow(2, attempt - 1), opts.maxDelayMs);
      const jitter = baseDelay * opts.jitterFactor * (Math.random() * 2 - 1);
      const delay = Math.max(0, baseDelay + jitter);

      log?.warn(
        { attempt, maxAttempts: opts.maxAttempts, delayMs: Math.round(delay), error: String(error) },
        'Request failed, retrying',
      );

      await new Promise((resolve) => setTimeout(resolve, delay));
    }
  }

  // Unreachable but TypeScript needs it
  throw new Error('Retry loop exited unexpectedly');
}
